web/packages/teleterm/src/services/appUpdater/autoUpdatesStatus.ts

```diff
@@ -0,0 +1,255 @@
+ * The version is selected using the following precedence:
+ * 1. `TELEPORT_TOOLS_VERSION` env var, if defined.
+ * 2. `tools_version` from the cluster manually selected to manage updates, if selected.
+ * 3. Most compatible version, if found.
+ * 4. If there's no version at this point, stop auto updates.
+ */
+export async function resolveAutoUpdatesStatus(sources: {
+  versionEnvVar: string;
+  managingClusterUri: string | undefined;
+  getClusterVersions(): Promise<GetClusterVersionsResponse>;
+}): Promise<AutoUpdatesStatus> {
+  if (sources.versionEnvVar === 'off') {
+    return { enabled: false, reason: 'disabled-by-env-var' };
+  }
+  if (sources.versionEnvVar) {
+    return {
+      enabled: true,
+      version: sources.versionEnvVar,
+      source: { kind: 'env-var' },
+    };
+  }
+
+  const clusterVersions = await sources.getClusterVersions();
+  return findManagingCluster({
+    clusterVersions,
+    managingClusterUri: sources.managingClusterUri,
+  });
+}
+
+/**
+ * Attempts to find a cluster managing updates, based on a user-selected cluster
+ * and all available clusters.
+ */
+function findManagingCluster(sources: {
+  managingClusterUri: string | undefined;
+  clusterVersions: GetClusterVersionsResponse;
+}): AutoUpdatesStatus {
+  const { reachableClusters, unreachableClusters } = sources.clusterVersions;
+  const candidateClusters = makeCandidateClusters(reachableClusters);
+  if (!candidateClusters.length) {
+    return {
+      enabled: false,
+      reason: 'no-cluster-with-auto-update',
+      unreachableClusters,
+    };
+  }
+
+  const clusterUri = sources.managingClusterUri;
+  if (clusterUri) {
+    const cluster = candidateClusters.find(c => c.clusterUri === clusterUri);
+    if (cluster) {
+      return {
+        enabled: true,
+        version: cluster.toolsVersion,
+        source: {
+          kind: 'managing-cluster',
+          candidateClusters,
+          clusterUri,
+          unreachableClusters,
+        },
+      };
+    }
+    logger.warn(
+      `Cluster ${clusterUri} managing updates was not found, continuing resolution.`
+    );
+  }
+
+  const mostCompatibleVersion =
+    findMostCompatibleToolsVersion(candidateClusters);
+  if (mostCompatibleVersion) {
+    return {
+      enabled: true,
+      version: mostCompatibleVersion,
+      source: {
+        kind: 'most-compatible',
+        candidateClusters,
+        unreachableClusters,
+        clustersUri: candidateClusters
+          .filter(c => c.toolsVersion === mostCompatibleVersion)
+          .map(c => c.clusterUri),
+      },
+    };
+  }
+
+  return {
+    enabled: false,
+    reason: 'no-compatible-version',
+    unreachableClusters,
+    candidateClusters,
+  };
+}
+
+/** Assigns each candidate a compatibility with client tools from other clusters. */
+function makeCandidateClusters(
+  versions: ClusterVersionInfo[]
+): CandidateCluster[] {
+  const candidates = versions.filter(v => v.toolsAutoUpdate);
```

Comment:
Within the context of `findMostCompatibleToolsVersion`, does this mean that if I have a cluster that doesn't have tool auto updates enabled, compatibility with this cluster is not going to be considered when choosing the final version?

Reply:
Right, I think the clusters with disabled auto updates should still be considered when choosing a compatible version, as otherwise we risk upgrading Connect to a version that's not compatible with the clusters that the user is connected to.